src/Components/Server/src/Circuits/CircuitIdFactory.cs

```diff
@@ -2,56 +2,114 @@
         // authenticate encrypt it using data protection.
         // For validation, the fact that we can unprotect the payload is guarantee enough.
-        public string CreateCircuitId()
+        public CircuitId CreateCircuitId()
         {
             var buffer = new byte[32];
             _generator.GetBytes(buffer);
             var payload = _protector.Protect(buffer);
 
-            return Base64UrlTextEncoder.Encode(payload);
+            return new CircuitId(buffer, GetHash(buffer), payload);
+        }
+
+        private static byte[] GetHash(byte[] buffer)
+        {
+            var hashAlgorithm = SHA256.Create();
+            var hash = hashAlgorithm.ComputeHash(buffer);
+            return hash;
+        }
+
+        public bool ValidateCircuitId(string circuitId, string cookie, out CircuitId id)
+        {
+            id = FromCookieValue(cookie);
+            var rawRequestId = Base64UrlTextEncoder.Decode(circuitId);
+            var requestTokenBytes = Base64UrlTextEncoder.Decode(id.RequestToken);
+            if (CryptographicOperations.FixedTimeEquals(rawRequestId, requestTokenBytes))
+            {
+                return true;
+            }
+            else
+            {
+                id = default;
+                return false;
+            }
         }
 
-        public bool ValidateCircuitId(string circuitId)
+        public bool ValidateCircuitId(string circuitId, ClaimsPrincipal user)
         {
             try
             {
-                var protectedBytes = Base64UrlTextEncoder.Decode(circuitId);
-                _protector.Unprotect(protectedBytes);
+                foreach (var claim in user.Claims)
+                {
+                    if (claim.Type.Equals(CircuitMatcherPolicy.IdClaimType))
+                    {
+                        var rawRequestId = Base64UrlTextEncoder.Decode(claim.Value);
+                        var requestTokenBytes = Base64UrlTextEncoder.Decode(circuitId);
 
-                // Its enough that we prove that we can unprotect the payload to validate the circuit id,
-                // as this demonstrates that it the id wasn't tampered with.
-                return true;
+                        if (CryptographicOperations.FixedTimeEquals(rawRequestId, requestTokenBytes))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                return false;
             }
             catch (Exception)
             {
                 // The payload format is not correct (either not base64urlencoded or not data protected)
                 return false;
             }
         }
+
+        internal CircuitId FromCookieValue(string value)
+        {
+            var payload = Base64UrlTextEncoder.Decode(value);
+            var id = _protector.Unprotect(payload);
+            return new CircuitId(id, GetHash(id), payload);
+        }
+    }
+
+    internal struct CircuitId
```

Comment:
Looks like this could be `readonly` for clarity

Reply:
Yeah, I also plan to do some cleanup here.